File: Part 1 - Data Preprocessing/data_preprocessing_template_endSection1.R
# Data Preprocessing

# Importing the dataset
dataset = read.csv('Data.csv')
  # Datasets start at index 1 in R

# Taking care of missing data
# Replace missing data in Age column
dataset$Age = ifelse(
                is.na(dataset$Age), # Condition
                ave(dataset$Age, FUN = function(x) mean(x, na.rm = TRUE)), # If condition is true
                dataset$Age # If condition is false
              )
#Replace missing data in Salary column
dataset$Salary = ifelse(
  is.na(dataset$Salary), # Condition
  ave(dataset$Salary, FUN = function(x) mean(x, na.rm = TRUE)), # If condition is true
  dataset$Salary # If condition is false
)

# Encoding categorical data
# Replace country names with numerical values
dataset$Country = factor(
                    dataset$Country, # Data that we want to transform into factors
                    levels = c('France', 'Spain', 'Germany'), # names of categories in our column
                    labels = c(1, 2, 3) # Labels - Which number to give to the levels, non-order related
                  )
# Replace purchased labels with numerical values
dataset$Purchased = factor(
  dataset$Purchased, # Data that we want to transform into factors
  levels = c('No', 'Yes'), # names of categories in our column
  labels = c(0, 1) # Labels - Which number to give to the levels, non-order related
)

# Splitting the dataset into the training set and test set
# install.packages('caTools')
library(caTools)
set.seed(123)
split = sample.split(dataset$Purchased, SplitRatio = 0.8)
training_set = subset(dataset, split == TRUE)
test_set = subset(dataset, split == FALSE)

# Feature scaling
training_set[, 2:3] = scale(training_set[, 2:3])
test_set[, 2:3] = scale(test_set[, 2:3])

File: Part 1 - Data Preprocessing/data_preprocessing_template_endSection1.py
# Data Preprocessing

# Importing the libraries
import numpy as np
import matplotlib.pyplot as plt
import pandas as pd

# Importing the dataset
dataset = pd.read_csv('Data.csv')
X = dataset.iloc[:, :-1].values
    # [all of the rows, all of the columns EXCEPT the last one]
Y = dataset.iloc[:, 3].values

# Taking care of missing data
np.set_printoptions(threshold = np.nan)
from sklearn.preprocessing import Imputer
# imputer = Imputer(missing_values = 'NaN', strategy = 'mean', axis = 0, verbose = 0, copy = True)
# CTRL+I to inspect the class
imputer = Imputer(missing_values = 'NaN', strategy = 'mean', axis = 0)
imputer = imputer.fit(X[:, 1:3]) # Lowerbound is inclusive, Upperbound is exclusive
X[:, 1:3] = imputer.transform(X[:, 1:3])

# Encoding categorical data
from sklearn.preprocessing import LabelEncoder, OneHotEncoder
labelEncoder_X = LabelEncoder()
X[:, 0] = labelEncoder_X.fit_transform(X[:, 0])
# Dummy encoding - X
oneHotEncoder = OneHotEncoder(categorical_features = [0])
X = oneHotEncoder.fit_transform(X).toarray() # Already specified the first column in the categorical_features
# Dummy encoding - Y
labelEncoder_Y = LabelEncoder()
Y = labelEncoder_Y.fit_transform(Y)

# Splitting the dataset into the training set and test set
from sklearn.cross_validation import train_test_split
X_train, X_test, Y_train, Y_test = train_test_split(X, Y, test_size = 0.2, random_state = 0)

# Feature scaling
from sklearn.preprocessing import StandardScaler
sc_X = StandardScaler()
X_train = sc_X.fit_transform(X_train)
X_test = sc_X.transform(X_test)